SOD.Common/Helpers/PluginDetection.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Linq;
+using BepInEx;
+using BepInEx.Configuration;
+using BepInEx.Unity.IL2CPP;
+using UniverseLib;
+
+namespace SOD.Common.Helpers
+{
+    public sealed class PluginDetection
+    {
+        /// <summary>
+        /// Raised right after the BepInEx chainloader finishes loading the
+        /// last plugin.
+        /// </summary>
+        public event EventHandler OnAllPluginsFinishedLoading;
+
+        /// <summary>
+        /// True if the BepInEx chainloader has finished loading all plugins,
+        /// false otherwise (is set to true after OnAllPluginsFinishedLoading).
+        /// </summary>
+        public bool AllPluginsFinishedLoading { get; private set; } = false;
+
+        private bool _initialized = false;
+
+        internal PluginDetection()
+        {
+            Initialize();
+        }
+
+        internal void Initialize()
+        {
+            if (_initialized)
+            {
+                return;
+            }
+            _initialized = true;
+
+            // Hook up the OnAllPluginsFinishedLoading event and related
+            var invokeEventAction = () => OnAllPluginsFinishedLoading?.Invoke(this, EventArgs.Empty);
+            IL2CPPChainloader.Instance.Finished += invokeEventAction;
+            IL2CPPChainloader.Instance.Finished += () => { AllPluginsFinishedLoading = true; };
+        }
+
+        /// <summary>
+        /// Get the full GUID of a plugin by searching for a partial GUID
+        /// across all loaded plugins.
+        /// </summary>
+        /// <remarks>
+        /// Should be called in response to Lib.BepInEx.OnAllPluginsFinishedLoading.
+        /// Useful for plugins such as Babbler, which use a prefix on their
+        /// GUID to influence load order.
+        /// </remarks>
+        /// <param name="partialPluginGuid">A partial GUID.</param>
+        /// <returns></returns>
+        /// <exception cref="System.InvalidOperationException"></exception>
+        public string GetPluginGuidFromPartialGuid(string partialPluginGuid)
+        {
+            var guids = IL2CPPChainloader.Instance.Plugins.Keys;
+            var matches = guids.Where(guid => guid.ToLower().Contains(partialPluginGuid.ToLower()));
```

Comment:
You probably want to ToArray() this to prevent multiple enumerations from the Count() and the First()

Reply:
ah I did not think of that, interesting. I'll change it